source/extensions/filters/network/common/redis/client_impl.cc

```diff
@@ -245,12 +247,28 @@ void ClientImpl::PendingRequest::cancel() {
   canceled_ = true;
 }
 
+void ClientImpl::initialize(const std::string& auth_password) {
+  if (!auth_password.empty()) {
+    // Send an AUTH command to the upstream server.
+    makeRequest(Utility::makeAuthCommand(auth_password), null_pool_callbacks);
+  }
+  // Any connection to replica requires the READONLY command in order to perform read.
+  // Also the READONLY command is a no-opt for the master.
+  // We only need to send the READONLY command iff it's possible that the host is a replica.
+  if (config_.readPolicy() != Common::Redis::Client::ReadPolicy::Master) {
+    makeRequest(Utility::ReadOnlyRequest::instance(), null_pool_callbacks);
+  }
+}
+
 ClientFactoryImpl ClientFactoryImpl::instance_;
 
 ClientPtr ClientFactoryImpl::create(Upstream::HostConstSharedPtr host,
-                                    Event::Dispatcher& dispatcher, const Config& config) {
-  return ClientImpl::create(host, dispatcher, EncoderPtr{new EncoderImpl()}, decoder_factory_,
-                            config);
+                                    Event::Dispatcher& dispatcher, const Config& config,
+                                    const std::string& auth_password) {
+  ClientPtr client =
+      ClientImpl::create(host, dispatcher, EncoderPtr{new EncoderImpl()}, decoder_factory_, config);
+  client->initialize(auth_password);
```

Comment:
Is there any reason to have an explicit initialize function vs. doing this from within the context of the create/constructor call?

Reply:
Having the explicit initialize function makes it easier to test and adhere to the single responsibility principal better.  The small perf hit of a function call is probably not a big issue given we don't create that many connections.  I'm open to change this if you think it's better to put it in the create call.

Reply:
It's fine if you need it for testing, but generally I try to avoid extra functions unless its really beneficial.